homeassistant/components/trafikverket_train/sensor.py

```diff
@@ -0,0 +1,172 @@
+    return next_weekday(today_date, WEEKDAYS.index(departure[0]))
+
+
+class TrainSensor(Entity):
+    """Contains data about a train depature."""
+
+    def __init__(self, train_api, name,
+                 from_station, to_station, weekday, time):
+        """Initialize the sensor."""
+        self._train_api = train_api
+        self._name = name
+        self._from_station = from_station
+        self._to_station = to_station
+        self._weekday = weekday
+        self._time = time
+        self._state = None
+        self._departure_state = None
+        self._delay_in_minutes = None
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Retrieve latest state."""
+        if self._time is not None:
+            departure_day = next_departuredate(self._weekday)
+            when = datetime.combine(departure_day, self._time)
+            try:
+                self._state = yield from \
+                    self._train_api.async_get_train_stop(
+                        self._from_station, self._to_station, when)
+            except ValueError as output_error:
+                _LOGGER.error("Departure %s encountered a problem: %s",
+                              when, output_error)
+        else:
+            when = datetime.now()
+            self._state = yield from \
+                self._train_api.async_get_next_train_stop(
+                    self._from_station, self._to_station, when)
+        self._departure_state = self._state.get_state().name
+        self._delay_in_minutes = self._state.get_delay_time()
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        if self._state is None:
+            return None
+        state = self._state
+        other_information = None
+        if state.other_information is not None:
+            other_information = ", ".join(state.other_information)
+        deviations = None
+        if state.deviations is not None:
+            deviations = ", ".join(state.deviations)
+        if self._delay_in_minutes is not None:
+            self._delay_in_minutes = \
+                self._delay_in_minutes.total_seconds() / 60
+        return {ATTR_CANCELED: state.canceled,
+                ATTR_DELAY_TIME: self._delay_in_minutes,
+                ATTR_PLANNED_TIME: state.advertised_time_at_location,
+                ATTR_ESTIMATED_TIME: state.estimated_time_at_location,
+                ATTR_ACTUAL_TIME: state.time_at_location,
+                ATTR_OTHER_INFORMATION: other_information,
+                ATTR_DEVIATIONS: deviations}
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon for the frontend."""
+        return ICON
+
+    @property
+    def state(self):
+        """Return the departure state."""
+        if self._state is not None:
+            return self._departure_state
```

Comment:
What type of data are we returning here? Is it a datetime string?

Reply:
States like this: on_time, delayed, canceled
Times are in attributes.
![image](https://user-images.githubusercontent.com/1937941/59860220-91099f00-937e-11e9-96bb-0dfe9ccdef24.png)

Reply:
You could consider changing state to planned time of departure and use device class timestamp. It will then show as a relative time until departure in the frontend.

What's the most useful state info out of all the attributes?

Reply:
Yes, that is a really neat feature.. I would say both values are equally important. The current setup will highlight if everything is ok or not,

Having time as state would need some additional logic to show the state since planned time is the timetable time and would not be valid to show as state if the train is delayed or canceled.

If the train is delayed and have an estimated time, this value will be in a separate attribute.
![image](https://user-images.githubusercontent.com/1937941/59869544-d389a680-9393-11e9-8247-ba11ac59205c.png)

Reply:
Yeah, timestamp state will be a bit more complex. I wanted to mention it at least.

Reply:
Timestamp is I think the way to go. A sensor that will always show the timestamp of the next time the train leaves.

Reply:
Hmm, yes, it is appealing at first. But given the fact how trains work in south of sweden there is most likely a delay on daily basis and then the planned time is no longer valid but the estimated time is the one to aim for. 
However, it is only an estimate and the train can arrive earlier in my experience, even though they often arrive later ;)

Reply:
Sweden 2.0. :laughing:

Reply:
If you set the time to a timestamp, users in the UI can choose how to show the time as they want. Time until it happens, just the time etc. It gives users the most flexibility.

Reply:
Looks neat in the UI with the time. 
I have added the device class as timestamp and moved the previous state to departure_state attribute.

   ```
 @property
    def state(self):
        """Return the departure state."""
        state = self._state
        if state is not None:
            if state.time_at_location is not None:
                return state.time_at_location
            elif state.estimated_time_at_location is not None:
                return state.estimated_time_at_location
            return state.advertised_time_at_location
        return None
```

![image](https://user-images.githubusercontent.com/1937941/59961804-d8a63d00-94dc-11e9-8033-19b5dd01c125.png)

And if the train is delayed it will show the estimated time..

![image](https://user-images.githubusercontent.com/1937941/59961837-d42e5400-94dd-11e9-893a-8f3ffd04e1d6.png)

However, if the departure is cancelled it will still show the planned time, not sure what the best way would be to show as state in that case.. Maybe that is up to the user, based on the attribute canceled and departure_state..

Reply:
Very nice! Yes, the user should decide what to do with the cancel attribute. We should keep the main state consistent in terms of type. It should always be a timepoint now.